### PR TITLE
Yet more inference/invalidation fixes

### PR DESCRIPTION
**Pull request by timholy**
This is the Pkg part of what I found in my investigations of invalidations due to extending `==`, c.f. https://github.com/JuliaLang/julia/pull/36282. This also requires https://github.com/JuliaLang/julia/pull/36280, so if that's not on nightly yet this wiil fail tests. (It also requires https://github.com/JuliaLang/julia/pull/36188, but I imagine that's on nightly by now.)

This will be easiest to review ignoring whitespace changes, due to some big `let` blocks from circumventing more cases of the infamous 15276.

### CODE REVIEW

File: src/Types.jl [hunk 1]
```diff
@@ -1240,13 +1244,13 @@ function find_registered!(ctx::Context,
...
         reg_abspath = abspath(registry.path)
         data = read_registry(joinpath(registry.path, "Registry.toml"))
-        for (_uuid, pkgdata) in data["packages"]
-              uuid = UUID(_uuid)
-              name = pkgdata["name"]
-              path = joinpath(reg_abspath, pkgdata["path"])
-              push!(get!(ctx.env.uuids, name, UUID[]), uuid)
-              push!(get!(ctx.env.paths, uuid, String[]), path)
-              push!(get!(ctx.env.names, uuid, String[]), name)
+        for (_uuid, pkgdata) in data["packages"]   # what types do these have?
```
Comment (timholy) [inline src/Types.jl:1247]: Anyone know what types `_uuid` and `pkgdata` are?

**Comment by timholy** [comment]
Ah, does Pkg get deployed to older Julia versions? Or just, sometimes developed on older Julia versions? Should I add version checks?

**Comment by KristofferC** [comment]
> Should I add version checks?

Depends on how much extra work there is. We might need to lose 1.4 and 1.5 compat soon anyway when we start using https://github.com/JuliaLang/julia/pull/36018.

File: src/Types.jl (hunk 1 above)
Comment (KristofferC) [reply src/Types.jl:1247]: `_uuid` should be `String` and `pkgdata` a `Dict{String, Any}` I think.

**Comment by timholy** [comment]
It wouldn't be much work, but your TOML-parser sounds like critical infrastructure to me. So perhaps this can just hang out until you're ready to commit to that and Julia 1.6.

File: src/Operations.jl [hunk 2]
```diff
@@ -1749,8 +1749,8 @@ function status(ctx::Context, pkgs::Vector{PackageSpec}=PackageSpec[];
         old_ctx = Context(;env=env_diff)
     end
     # display
-    filter_uuids = [pkg.uuid for pkg in pkgs if pkg.uuid !== nothing]
-    filter_names = [pkg.name for pkg in pkgs if pkg.name !== nothing]
+    filter_uuids = [pkg.uuid::UUID for pkg in pkgs if pkg.uuid !== nothing]
```
Comment (KristofferC) [inline src/Operations.jl:1752]:
```suggestion
    filter_uuids = UUID[pkg.uuid for pkg in pkgs if pkg.uuid !== nothing]
```

maybe? Same below.

**Comment by KristofferC** [comment]
I think we should drop 1.4 and merge this. We should probably then add a matrix entry to the Travis runner so that we also run with `JULIA_PKG_SERVER=""` to make sure that it doesn't regress. Also, we can probably get rid of AV in favor of Windows on Travis.

File: src/Operations.jl (hunk 2 above)
Comment (timholy) [reply src/Operations.jl:1752]: The version here avoids one runtime dispatch and one invalidation-worthy call to `convert` (from inside the `setindex!` call):
```julia
julia> f(x) = Int[x]
f (generic function with 1 method)

julia> g(x) = [x::Int]
g (generic function with 1 method)

julia> code_typed(f, (Integer,))
1-element Array{Any,1}:
 CodeInfo(
1 ─ %1 = $(Expr(:foreigncall, :(:jl_alloc_array_1d), Array{Int64,1}, svec(Any, Int64), 0, :(:ccall), Array{Int64,1}, 1, 1))::Array{Int64,1}
│        Base.setindex!(%1, x, 1)::Any
└──      return %1
) => Array{Int64,1}

julia> code_typed(g, (Integer,))
1-element Array{Any,1}:
 CodeInfo(
1 ─      Core.typeassert(x, Main.Int)::Int64
│   %2 = π (x, Int64)
│   %3 = $(Expr(:foreigncall, :(:jl_alloc_array_1d), Array{Int64,1}, svec(Any, Int64), 0, :(:ccall), Array{Int64,1}, 1, 1))::Array{Int64,1}
│        Base.arraysize(%3, 1)::Int64
└──      goto #3 if not true
2 ─      Base.arrayset(false, %3, %2, 1)::Array{Int64,1}
3 ┄      goto #4
4 ─      return %3
) => Array{Int64,1}
```

Comment (KristofferC) [reply src/Operations.jl:1752]: Alright, thanks for exaplining.

**Comment by KristofferC** [comment]
I'll push to this branch for the stuff I mentioned in the last comment

**Comment by timholy** [comment]
Are you fixing the conflict or shall I? I don't want to cause more trouble by trying to be helpful.

**Comment by timholy** [comment]
Looks like this should pass. I therefore took the liberty of reorganizing the commits into more logical form. Feel free to merge or squash, whatever you prefer.

**Comment by timholy** [comment]
Looks good to go. Might want to increase the amount of time on windows, since one of the two jobs timed out and the other was fairly close to the limit.

**Comment by KristofferC** [comment]
I'll put back AV then, it seems faster there.

**Comment by timholy** [comment]
BTW, might be worth mentioning that after I reconfigured the commits, I'm pretty sure the only two that need `nightly` are the `unique` and `maximum` commits. Let me know if you want me to do anything with that observation.

**Pull request merged**